second/8-flowapp.py

```diff
@@ -0,0 +1,245 @@
+# ライブラリのインポート
+import copy
+
+import cv2
+import numpy as np
+
+# Opticalflowを元にどうボールを動かすか(Version1 or Version2)
+# 移動度が閾値(threshold)を超えた場合のみボールを動かす(目安 : Version1→30, Version2→100)
+# ボールを一回に何ピクセル動かすか(move_distance)
+# 詳しくはスライドの(補足 : オプティカルフローをどう判断するか?)を参照
+flow_usage = "Version1"
+threshold = 30
+move_distance = 10
+
+# Webカメラ設定
+cap = cv2.VideoCapture(0)
+
+# Shi-Tomasiのコーナー検出パラメータ
+feature_params = dict(
+    maxCorners=100,  # 保持するコーナー数, int型
+    qualityLevel=0.3,  # 最良値(最大固有値の割合?), float型
+    minDistance=7,  # この距離内のコーナーを棄却, float型
+    blockSize=7,
+)  # 使用する近傍領域のサイズ, int
+
+# Lucas-Kanade法のパラメータ
+lk_params = dict(
+    winSize=(15, 15),  # 検索ウィンドウのサイズ
+    maxLevel=2,  # 検出器の構造を決める(そのままで良い)
+    criteria=(cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT, 10, 0.03),
+)  # 検索終了条件
+
+# ランダムに色を１００個生成（値0～255の範囲で100行3列のランダムなndarrayを生成）
+color = np.random.randint(0, 255, (100, 3))
+
+# アプリ用のスタジアム、ボール画像を読みこみ
+ret, frame = cap.read()
+ball_img = cv2.imread("./image_data/ball.png")
+stadium_img = cv2.imread("./image_data/stadium.png")
+
+# Webカメラの画面の大きさにスタジアムを合わせる
+stadium_img = cv2.resize(stadium_img, (frame.shape[1], frame.shape[0]))
+
+# ボールの高さ、幅の[半分](半分だから注意！！)
+# (注意!)今回ボールの大きさが H:198、W:200と両方偶数のためこれで良いが、奇数の場合は工夫が必要
+ball_h, ball_w = ball_img.shape[0] // 2, ball_img.shape[1] // 2
+
+# ボールの初期位置（中心座標)をスタジアムの中心に設定
+idx_h = stadium_img.shape[0] // 2
+idx_w = stadium_img.shape[1] // 2
+
+# はじめボールは中央に配置
+stadium = copy.deepcopy(stadium_img)
+stadium[
+    (idx_h - ball_h) : (idx_h + ball_h), (idx_w - ball_w) : (idx_w + ball_w)
+] = ball_img
+
+# フレームカウント
+count = 0
+
+# 実行
+while True:
+
+    ret, frame = cap.read()
+
+    # 10フレームに一回特徴点を更新
+    if count % 10 == 0:
+        # グレースケールに変換
+        gray_first = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+        feature_first = cv2.goodFeaturesToTrack(gray_first, mask=None, **feature_params)
+        flow_mask = np.zeros_like(frame)  # フロー書き出し用の画像更新
+        # カウントの初期化
+        count = 0
+
+    # 残りの9回はオプティカルフローを計算してボールを動かす
+    else:
+        # グレースケールに変換
+        gray_next = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+
+        # オプティカルフロー検出
+        feature_next, status, err = cv2.calcOpticalFlowPyrLK(
+            gray_first, gray_next, feature_first, None, **lk_params
+        )
+
+        # オプティカルフローを検出した特徴点を選別（0：検出せず、1：検出した）
+        good_first = feature_first[status == 1]
+        good_next = feature_next[status == 1]
+
+
+        """
+        フローの結果をどのように利用するか(flow_usage)で分岐
+        Version1 → 最大の移動度を持つ特徴点を基にボールを移動
+        Version2 → 全特徴点の移動度の合計を基にボールを移動
+        """
+
```

Comment:
`flow_usage`が，`Version1`にも`Version2`にもどちらにも当てはまらないときの警告があってもいいかも
```suggestion
        assert flow_usage in ["Version1", "Version2"]
```